### PR TITLE
Minimize configs

### PR DESCRIPTION
**Pull request by rekmarks**
This PR adds one final responsibility to our config validation script: to ensure that our configs are _minimal_.

This means that our configs do not contain any "useless" rules, meaning:

- Rules that are disabled but never enabled by an extended config.
- Rules that are configured identically by the package's extended configs.

For the purpose of determining the "usefulness" of rules, we include our base config (`@metamask/eslint-config`) in the set of extended configs, since it should always be extended by the consumer in practice.

### Why?

This is definitely a stretch goal, but I think config minimization is worth it for two reasons:

1. It's DRY.
2. We will never again ask if this or that rule is included because it's recommended by some authority, or because _we wanted it_.

This should make our configs easier to review and maintain, as well as better expressions of our stylistic preferences.

### Note for Reviewers

You'll note that some of the rule snapshot files have changed, most notably the TypeScript ones. In _all_ cases, for _all_ snapshots, all of the removed rules were set to `off`, _except_ `@typescript-eslint/quotes`. Our script did not complain about that rule, but it turned to be uselessly configured. See [the Prettier docs](https://github.com/prettier/eslint-config-prettier/#quotes) to understand why.

### CODE REVIEW

File: scripts/validate-configs.js [hunk 1]
```diff
@@ -272,17 +285,64 @@ function getMetamaskConfigs() {
...
+ * @returns {{ extended: Record<string, unknown>, full: Record<string, unknown>}} An
+ * object containing the extended and full flat rules.
+ */
+function getExtendedAndFullFlatRules(packageName, config) {
+  const isBaseConfig = packageName === BASE_CONFIG_NAME;
+  const flatConfig = isBaseConfig
+    ? getFlatConfig(config)
+    : getFlatConfigWithBaseConfig(config);
+
+  const extendedFlatRules = getFlatRules(flatConfig.slice(0, -1));
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply scripts/validate-configs.js:323]: Everything _but_ the last entry. The last entry is our config. I'll add a comment explaining this.

[the remark this answers is not in the document]
Comment (Gudahtt) [reply scripts/validate-configs.js:323]: D'oh! Of course. OK, that makes sense.

File: scripts/validate-configs.js [hunk 2]
```diff
@@ -88,47 +111,94 @@ async function main() {
...
+ *
+ * @param {Record<string, string[]>} - A map of package names to arrays with
+ * violated rules, if any.
+ * @returns {boolean} Whether the given map contains any violations.
  */
-function tabs(numTabs) {
-  const TAB = '    ';
-  return TAB + new Array(numTabs).join(TAB);
+function hasViolations(violationsMap) {
+  for (const violations of Object.values(violationsMap)) {
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply scripts/validate-configs.js]: `.some()` is as understandable _and_ more concise, however! e854cd0

File: scripts/validate-configs.js [hunk 3]
```diff
@@ -272,17 +285,64 @@ function getMetamaskConfigs() {
...
+  const isBaseConfig = packageName === BASE_CONFIG_NAME;
+  const flatConfig = isBaseConfig
+    ? getFlatConfig(config)
+    : getFlatConfigWithBaseConfig(config);
+
+  const extendedFlatRules = getFlatRules(flatConfig.slice(0, -1));
+
+  return {
+    extended: extendedFlatRules,
+    full: getFlatRules(isBaseConfig ? flatConfig : getFlatConfig(config)),
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply scripts/validate-configs.js]: Yeah, done here: e854cd0

Also, _I_ was confused by the use of `full`, so I renamed it to `own` and added explanatory comments.

File: scripts/validate-configs.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (rekmarks) [reply scripts/validate-configs.js:328]: e854cd0

File: scripts/validate-configs.js [hunk 4]
```diff
@@ -381,3 +450,75 @@ function populateRecommendedRules(configArray) {
...
+ * string describing them.
+ *
+ * @param {Record<string, string[]>} violationsMap - A map of config package
+ * names to rules flagged as violations.
+ * @returns {string} A formatted string listing the violations.
+ */
+function getViolationsString(violationsMap) {
+  let str = '';
+  Object.entries(violationsMap).forEach(([packageName, violatedRules]) => {
+    if (violationsMap[packageName].length > 0) {
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply scripts/validate-configs.js]: e854cd0

File: scripts/validate-configs.js [hunk 5]
```diff
@@ -20,59 +25,77 @@ const RULES_SNAPSHOT_PATH = 'rules-snapshot.json';
...
  * computed snapshot to the snapshot stored on disk, and exit with an error if
  * they aren't equal.
  */
 async function main() {
   const metamaskConfigs = getMetamaskConfigs();
   const requiredPrettierRules = getRequiredPrettierRules();
 
-  const prettierViolations = getConfigViolationsMap(metamaskConfigs);
+  const prettierViolations = getViolationsMap(metamaskConfigs);
+  const minimalismViolations = getViolationsMap(metamaskConfigs);
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply scripts/validate-configs.js:59]: 😅 

I don't think refactoring is worth it at this point, but I at least added some explanatory comments: 3ce7871

File: scripts/validate-configs.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply scripts/validate-configs.js]: Hah, I was about to comment about the name being confusing too! This is _much_ clearer - fantastic improvement

File: scripts/validate-configs.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply scripts/validate-configs.js:59]: Thanks! That helps

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Pull request merged**